### DISCUSSION TITLE
form conversion

### DISCUSSION THREAD

**Discussion by coenttb** [pointfreeco/swift-parsing#215] [Ideas]
Dear community,

The PointFreeRouter library in the pointfreeco repo contains a Conversion.form. This is useful to send FormData in codable form to the server and decode it there for further use.

I was wondering if there is any reason why this is not included in the swift-parsing library too? Considering the Conversion.json is here?

**Comment by coenttb** [comment]
So I found out there is an idiomatic way to do this. Example follows:
```
public struct SubmitData:Codable, Hashable {
    let stringField: String
    let boolField:Bool

    enum CodingKeys:String, CodingKey {
        case stringField
        case boolField
    }
}
Route(.case(SiteRoute.FormPage.post)) {
        Method.post
        Path { "submit" }
        Body {
            FormData {
                Parse(.memberwise(SubmitData.init)) {
                    Field(SubmitData.CodingKeys.stringField.rawValue, default: "")
                    Field(SubmitData.CodingKeys.boolField.rawValue, default: false) {
                        Bool.parser()
                    }
                }
            }
        }
    }
```

I also found some weirdness, where the value of the bool input (checkbox) needs to be "true" and the Field parsers's default need to be false. Otherwise it won't encode/decode the boolean. I'll leave it to you to figure out why.

**Reply by stephencelis** [reply]
@tenthijeboonkkamp Without a default, you need an `Optionally` to wrap the `Field`, since it will fail if the field is omitted (and a checkbox is omitted when submitted and unchecked).
